src/Agent/NewRelic/Agent/Core/Agent.cs

```diff
@@ -562,19 +567,82 @@ internal void Detach(bool removeAsync, bool removePrimary)
+            // Specificity is just the length of the string, minus any wildcard character at the end
+            // Apply rules in order. Include rules add attributes to filteredContextData from unfilteredContextData, and exclude rules remove attributes from filteredContextData
+
+            var allRulesInOrder = _orderedCludeRuleList ?? (_orderedCludeRuleList = GetOrderedCludeRuleList()); 
+
+            if (_includeRuleList.Count() == 0) // empty "include" list, so include everything
             {
                 filteredContextData = unfilteredContextData;
             }
-            else
+
+            // Now apply the rules in order
+            // TODO: maybe optimize by just working with keys as lists of strings until it comes time to return the actual data
+            foreach (var rule in allRulesInOrder)
             {
-                filteredContextData = unfilteredContextData.Where(property => _configurationService.Configuration.ContextDataInclude.Contains(property.Key)).ToDictionary();
+                if (rule.Include)
+                {
+                    filteredContextData = filteredContextData.Concat(unfilteredContextData.Where(x => Regex.IsMatch(x.Key, rule.Text))).ToDictionary();
+                }
+                else
+                {
+                    filteredContextData = filteredContextData.Where(x => !Regex.IsMatch(x.Key, rule.Text)).ToDictionary();
+                }
             }
-            // now handle the exclude list - excludes win over includes per the spec
-            return filteredContextData.Where(property => !_configurationService.Configuration.ContextDataExclude.Contains(property.Key)).ToDictionary();
+            return filteredContextData;
+        }
+
+        private List<LogContextDataFilterRule> GetCludeRulesFromConfig(bool isIncludeRules)
+        {
+            List<LogContextDataFilterRule> rules = new List<LogContextDataFilterRule>();
+            var configRuleList = isIncludeRules ? _configurationService.Configuration.ContextDataInclude : _configurationService.Configuration.ContextDataExclude;
+            foreach (var configRule in configRuleList)
+            {
+                rules.Add(new LogContextDataFilterRule(configRule, isIncludeRules));
+            }
+            return rules;
+        }
+
+        private List<LogContextDataFilterRule> GetOrderedCludeRuleList()
+        {
+            var includeRuleList = _includeRuleList ?? (_includeRuleList = GetCludeRulesFromConfig(true));
+            var excludeRuleList = _excludeRuleList ?? (_excludeRuleList = GetCludeRulesFromConfig(false));
+
+            // This should work because OrderBy uses a stable sort.  By sorting each list individually,
+            // concactating them, and then sorting again, we get a list of rules ordered by length (specificity),
+            // with includes before excludes for each level of specificity
+            return includeRuleList.OrderBy(rule => rule.Specificity).Concat(
```

Comment:
This feels a bit clunky to me. While we probably won't have many include/exclude items, this requires three sorts and relies on (perhaps) non-intuitive behavior of the `OrderBy` function. Maybe there's a simpler way to do this?